core/worker.go

```diff
@@ -614,6 +622,173 @@ func (w *worker) GeneratePendingHeader(block *types.WorkObject, fill bool) (*typ
+func (w *worker) transactionOrderingLoop() {
+	defer func() {
+		if r := recover(); r != nil {
+			w.logger.WithFields(log.Fields{
+				"error":      r,
+				"stacktrace": string(debug.Stack()),
+			}).Error("Go-Quai Panicked")
+		}
+	}()
+	defer w.wg.Done()
+
+	for {
+		select {
+		case orderInfo := <-w.orderTransactionCh:
+			w.OrderTransactionSet(orderInfo.txs, orderInfo.gasUsedAfterTransaction, orderInfo.block)
+		case <-w.exitCh:
+			return
+		}
+	}
+}
+
+// OrderTransactionSet takes in the set of transactions and
+// gasUsedAfterTransaction, picks subset of the transactions or change the order
+// of transactions as to increase the revenue of the miner
+// 1) First filtering out the non etx transactions
+// 2) Create a separate unique gas price list
+// 3) Iterate through the gas price list and for each iteration, add all the
+// transactions that are above the gas price and calculate the revenue, it the
+// revenue if greater than the previous best revenue, overwrite the best
+// transaction set, once the whole gas price list is exhausted, there will be a
+// set which maximizes the revenue
+// 4) Call the GeneratePendingHeader from inside and supply the best transaction
+// set
+
+func (w *worker) OrderTransactionSet(txs []*types.Transaction, gasUsedAfterTransaction []uint64, block *types.WorkObject) {
+	defer func() {
+		if r := recover(); r != nil {
+			w.logger.WithFields(log.Fields{
+				"error":      r,
+				"stacktrace": string(debug.Stack()),
+			}).Error("Go-Quai Panicked")
+		}
+	}()
+	type TransactionInfo struct {
+		Tx       *types.Transaction
+		GasPrice *big.Int
+		GasUsed  uint64
+	}
+
+	if len(txs) != len(gasUsedAfterTransaction) {
+		w.logger.Warnf("len of txs %d is not equal to the len of the gas used transactions %d", len(txs), len(gasUsedAfterTransaction))
+		return
+	}
+
+	// Extract the non external transactions from the txs list
+	nonExternalTxs := make([]*types.Transaction, 0)
+	nonExternalGasUsed := make([]uint64, 0)
+	for i, tx := range txs {
+		if tx.Type() != types.ExternalTxType {
+			nonExternalTxs = append(nonExternalTxs, tx)
+			var gasUsed uint64
+			if i == 0 {
+				gasUsed = gasUsedAfterTransaction[i]
+			} else {
+				gasUsed = gasUsedAfterTransaction[i] - gasUsedAfterTransaction[i-1]
+			}
+			nonExternalGasUsed = append(nonExternalGasUsed, gasUsed)
+		}
+	}
+	if len(nonExternalTxs) == 0 {
+		return
+	}
+
+	// gas price and the gas used is extracted from each of these non external
+	// transactions for the sorting
+	txInfos := make([]TransactionInfo, 0)
+	for i, tx := range nonExternalTxs {
+		// If there is a Qi transcation, try to get the fee from the pool
+		// If it doesnt exist in the pool, we will ignore the transaction
+		// because, its expensive to calculate the fee
+		var gasPrice *big.Int
+		if tx.Type() == types.QiTxType {
+			fee, exists := w.txPool.qiTxFees.Peek([16]byte(tx.Hash().Bytes()))
```

Comment:
The txs you get from QiPoolPending (and TransactionsByPriceAndNonce) already have the fee attached to them, you just have to use that. I guess you can also do it this way, it's just less efficient

Reply:
The Qi fee is in Qi in either case so needs to be converted to Quai

Reply:
Yeah, there were issues while testing with Qi, fixed them all now, take a look again

Reply:
I don't see a difference here

Reply:
This is the ordering transactions scenario, i convert this fee to the Quai fee after this, and thats what i use to make the list

Reply:
I believe this issue is solved now, so resolving this comment @jdowning100, we can make an issue if there is any